File: 2021_10.py
import sys;
import logging;

AOC_DAY=10
AOC_YEAR=2021

chunk_characters = {
        '(':')',
        '[':']',
        '{':'}',
        '<':'>',
}
illegal_character_score = {
        ')': 3,
        ']': 57,
        '}': 1197,
        '>': 25137,
}
complete_character_score = {
        ')': 1,
        ']': 2,
        '}': 3,
        '>': 4,
};

def is_corrupted(line, logger):
    # return (score of corrupted line; score of incomplete line)
    # incomplete_score: of corrupted line; 0 if valid
    # complete_score: of incomplete line
    match = [];
    for ch in line:
        if ch in chunk_characters:
            match.append(chunk_characters[ch]);
        elif ch in illegal_character_score:
            try:
                p = match.pop();
            except:
                p = '';
            if(p != ch):
                logger.debug(" Found illegal character {} (should have" \
                        " matched {})! Returning {}.".format(ch, p,
                            illegal_character_score[ch]));
                return (illegal_character_score[ch], 0);
        else:
            raise Exception("Unrecognized character {}!".format(ch));

    print("---------- match {}".format(match));
    incomplete_score = 0;
    match.reverse();
    for ch in match:
        print("-- ch adds {} to {}".format(ch, incomplete_score));
        incomplete_score = incomplete_score * 5 + complete_character_score[ch];
    return (0, incomplete_score);


if('__main__' == __name__):
    import argparse;

    # deal with command line arguments
    parser = argparse.ArgumentParser(
            description = 'Advent of Code {} day {} solution'.format(
                AOC_YEAR, AOC_DAY));
    logger = logging.Logger('AOC{}d{}'.format(AOC_YEAR, AOC_DAY));
    logger.addHandler(logging.StreamHandler());
    logger.setLevel(logging.INFO);

    parser.add_argument(
        '-2', '--part2', action='store_true',
        help='part 2?', default=False);
    parser.add_argument(
        '-v', '--verbose', action='store_true',
        help='be verbose', default=False);

    args = parser.parse_args();
    if(args.verbose):
        logger.setLevel(logging.DEBUG);

    illegal_score = 0;
    incomplete_score = [];
    line = sys.stdin.readline().strip();
    while len(line) > 0:
        logger.debug("parsing line {}".format(line));
        (illegal_line_score, incomplete_line_score) = is_corrupted(line, logger)
        illegal_score += illegal_line_score;
        if(0 == illegal_line_score):
            incomplete_score.append(incomplete_line_score);
        line = sys.stdin.readline().strip();

    print(illegal_score);
    incomplete_score.sort();
    print(incomplete_score);
    print(incomplete_score[int(len(incomplete_score)/2)]);



# guesses: 5849401116 (too high)

File: 2021_10.rb
require 'logger'

AOCDAY = 10
AOCYEAR = 2021

CorruptedCharacterScores = {
  ')' => 3,
  ']' => 57,
  '}' => 1197,
  '>' => 25137,
}
CharacterMatch = {
  ')' => '(',
  ']' => '[',
  '}' => '{',
  '>' => '<',
}
IncompleteCharacterScores = {
  '(' => 1,
  '[' => 2,
  '{' => 3,
  '<' => 4,
}

def score_incomplete(charstack)
  score = 0
  charstack.reverse.each do |ch|
    score *= 5
    score += IncompleteCharacterScores[ch]
  end
  score
end

def run(inputfile, logger, **kwargs)
  incomplete_scores = []

  # part 1: find corrupted lines
  score = 0;
  linenumber = 0
  File.open(inputfile).each_line do |line|
    logger.debug("Line: #{line.strip}");
    linenumber += 1
    charstack = []
    line.strip.each_char do |ch|
      matchchar = CharacterMatch[ch]
      logger.debug("  char #{ch} matchchar #{matchchar}");
      if matchchar.nil?
        # no matching character. Push this character on to the stack.
        charstack << ch
      else
        ch_to_match = charstack.pop
        logger.debug("char #{ch} matchchar #{matchchar} ch_to_match #{ch_to_match}");
        if matchchar != ch_to_match
          # syntax error. Score, and move to the next line.
          logger.debug("Found mismatched character '#{ch}' on line #{linenumber}");
          score += CorruptedCharacterScores[ch]
          charstack = []
          break
        end
      end
    end

    # if we've processed every character but there remain characters in the
    # stack, this line is incomplete. 
    if charstack.length > 0
      logger.debug("Found incomplete line #{linenumber}. Remaining to match: #{charstack.join('')}");
      incomplete_scores << score_incomplete(charstack)
    end
  end

  puts("Syntax error score: #{score}");

  logger.debug("Incomplete lines scores: #{incomplete_scores} (found #{incomplete_scores.length})")
  puts("Incomplete lines score: #{incomplete_scores.sort[incomplete_scores.length.div(2)]}")
end


if __FILE__ == $0
  require 'optparse'

  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: $0 [options] <inputfile> [<more inputfiles>]\n" \
      "  Advent of Code #{AOCYEAR} day #{AOCDAY}"

    opts.on("-v", "--verbose", "Be verbose") do |v|
      options[:verbose] = v
    end

    opts.on("-2", "--part2", "Run part 2") do |part2|
      options[:part2] = true
    end

    opts.on("-h", "--help", "Print this help message") do
      puts opts
      exit
    end
  end.parse!

  logger = Logger.new(STDOUT);
  logger.level = (options.include? :verbose) ? Logger::DEBUG : Logger::WARN

  ARGV.each do |inputfile|
    run(inputfile, logger, **options)
  end
end
